cluster_client.cpp

```diff
@@ -282,15 +320,29 @@ bool cluster_client::get_key_for_conn(unsigned int conn_id, int iter, unsigned l
         *key_index = m_obj_gen->get_key_index(iter);
         m_key_len = snprintf(m_key_buffer, sizeof(m_key_buffer)-1, "%s%llu", m_obj_gen->get_key_prefix(), *key_index);
 
-        // check if the key match for this connection
         unsigned int hslot = calc_hslot_crc16_cluster(m_key_buffer, m_key_len);
+
+        // check if the key match for this connection
         if (m_slot_to_shard[hslot] == conn_id) {
             m_reqs_generated++;
             return true;
         }
 
+        // handle key for other connection
+        unsigned int other_conn_id = m_slot_to_shard[hslot];
+
+        // in case we generated key for connection that is disconnected, 'slot to shard' map may need to be updated
```

Comment:
@YaacovHazan It's not clear why would this happen and what should happen next.

Reply:
This scenario related to how our cluster and DMC handle the topology in cluster mode.
when we migrate shard from node to node it's take some time to all DMC's to get this information.

consider this scenario:
topology
node1 with shard1 0-8192
node2 with shard2 8193-16384

memtier client run and open two connections (conn1 & conn2)
in some point we migrate shard1 to node2.
conn1 will get moved from DMC1 and will send CLUSTER_SLOT to DMC1 that will send the new topolgy:
node2 0-16384

but since DMC2 didn't receive yet the new topolgy, when node2 will send key that map to 0-8192,
he will get MOVED and will send CLUSTER_SLOT to DMC2 that will send the old topolgy (DMC2 didn't updated yet):
node1 with shard1 0-8192
node2 with shard2 8193-16384

in this point conn1 may not connected any more, since we have no shards in node1 and we removing the endpoint from node1.
so node2 will send only keys that map to 8193-16384 with no MOVED reply, which is correct, but he will not know that he can and need to send also keys to 0-8192.

for that case I added this handle of situation that we generate key to connection that is disconnected, but I'm not sure we want to add this internal handling explanation in the comment